### PR TITLE
LPS-149390 Uses getCanonicalURL to get an url without query string like we are doing in SocialBookmarksTag

### PR DESCRIPTION
**Pull request by ealonso**

### CODE REVIEW

**Comment by ealonso** [comment]
Hey @georgel-pop-lr,

Could you please take a look at it and let me know if it fix the issue? I already tested it and it seems to be working fine for me, I just wanna double check it.

Thanks!!

**Comment by georgel-pop-lr** [comment]
> Hey @georgel-pop-lr,
> 
> Could you please take a look at it and let me know if it fix the issue? I already tested it and it seems to be working fine for me, I just wanna double check it.
> 
> Thanks!!

With these changes, we are losing the friendlyURL when available, as viewURL contains it, and when is null then we are getting the viewFullContentURL.toString() wich returns the internal one (not friendly ...es/7/-/asset_publisher/EHzmpENo98X8/content/web-content-1-es-).

The problem is that sometimes viewURL is friendlyURL and sometimes is not (is http://localhost:8080/es/7/-/asset_publisher/EHzmpENo98X8/content/web-content-1-es-), so is hard to detect when is not friendly and you need to remove params... 

Regards.

File: modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/util/AssetPublisherHelperImpl.java [hunk 1]
```diff
@@ -520,8 +524,11 @@ public String getAssetSocialURL(
 				viewFullContentURL.toString());
 
 			if (Validator.isNotNull(viewURL)) {
-				return viewURL;
+				viewURL = viewFullContentURL.toString();
```
Comment (georgel-pop-lr) [inline modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/util/AssetPublisherHelperImpl.java:527]: This removes friendlyURL completely :(

Comment (ealonso) [reply modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/util/AssetPublisherHelperImpl.java:527]: My bad

**Comment by ealonso** [comment]
Hey @georgel-pop-lr,

Could you please let me know the steps to reproduce the cases were the viewURL is friendlyURL and the cases were is not friendlyURL?

Regards,
Eudaldo

**Comment by georgel-pop-lr** [comment]
> Hey @georgel-pop-lr,
> 
> Could you please let me know the steps to reproduce the cases were the viewURL is friendlyURL and the cases were is not friendlyURL?
> 
> Regards, Eudaldo

Hi,

With the latest change, we go back to my first comment when we add layout ID in front of friendlyURL, and it breaks it.

This because viewURL is never null, at least on my tests, and if you have a friendlyURL we don't need to use getCanonical function on it.

viewURL variable is friendly when there is a Display page set up, if not, it is not friendly and has a long param.

For the moment, the param seems to only occurs for not friendly URL.

Also, if we apply getCanonicalURL we need to apply again the getAlternateURL, otherwise we lose the language in the URL. This language already exists in the viewURL, but is getting removed by the getCanonicalURL function. Normally would be enough to only remove the long param.

Hope it helps.

Regards.

**Comment by ealonso** [comment]
Hey @georgel-pop-lr,

Could you please let me know the steps where we add the layout ID in front of friendlyURL?

Regards,
Eudaldo

**Comment by georgel-pop-lr** [comment]
> Hey @georgel-pop-lr,
> 
> Could you please let me know the steps where we add the layout ID in front of friendlyURL?
> 
> Regards, Eudaldo

Hi @ealonso,

It is added by this line https://github.com/liferay/liferay-portal/blob/aff2303da68a3f7b9b732755eeb149e70eb1ed7b/portal-impl/src/com/liferay/portal/util/PortalImpl.java#L1454 

And seems to come from variable 	canonicalLayoutFriendlyURL = defaultLayoutFriendlyURL; https://github.com/liferay/liferay-portal/blob/aff2303da68a3f7b9b732755eeb149e70eb1ed7b/portal-impl/src/com/liferay/portal/util/PortalImpl.java#L1407 if this condition is true https://github.com/liferay/liferay-portal/blob/aff2303da68a3f7b9b732755eeb149e70eb1ed7b/portal-impl/src/com/liferay/portal/util/PortalImpl.java#L1398

Hope it helps.

Regards.

**Comment by ealonso** [comment]
ci:forward

**Comment by ealonso** [comment]
Hey @georgel-pop-lr,
I made some changes in your pr to adapt it to the portal patterns, instead using 
```
StringUtil.split(viewURL, "?")[0]
```
I changed the code to use, since we are using it in other places like ServletPaths#113:
```
int index = url.indexOf(CharPool.QUESTION);

if (index != -1) {
	url = url.substring(0, index);
}
```

**Comment by georgel-pop-lr** [comment]
> Hey @georgel-pop-lr, I made some changes in your pr to adapt it to the portal patterns, instead using
> 
> ```
> StringUtil.split(viewURL, "?")[0]
> ```
> 
> I changed the code to use, since we are using it in other places like ServletPaths#113:
> 
> ```
> int index = url.indexOf(CharPool.QUESTION);
> 
> if (index != -1) {
> 	url = url.substring(0, index);
> }
> ```

Looks good to me, thanks.

I will test it today.

**Comment by ealonso** [comment]
ci:forward

**Comment by ealonso** [comment]
ci:test:relevant

**Comment by ealonso** [comment]
ci:test:relevant

**Comment by ealonso** [comment]
ci:forward

**Comment by Tim-Cao** [comment]
ci:forward

**Comment by Tim-Cao** [comment]
See https://github.com/brianchandotcom/liferay-portal/pull/115176

**Comment by Tim-Cao** [comment]
ci:close

**Pull request closed without merging**